internal/output/html.go

```diff
@@ -0,0 +1,475 @@
+	for _, group := range vulnPkg.Groups {
+		slices.SortFunc(group.IDs, identifiers.IDSortFunc)
+		representID := group.IDs[0]
+		groupIDs[representID] = group
+
+		if !group.IsCalled() {
+			uncalledVulnIDs[representID] = true
+			uncalledPackages[vulnPkg.Package.Name] = true
+		} else {
+			calledPackages[vulnPkg.Package.Name] = true
+		}
+	}
+}
+
+// buildHTMLResult builds the final HTMLResult object from the ecosystem map and total vulnerability count.
+func buildHTMLResult(ecosystemMap map[string][]HTMLSourceResult, resultCount HTMLVulnCount) HTMLResult {
+	var ecosystemResults []HTMLEcosystemResult
+	var osResults []HTMLEcosystemResult
+	for ecosystem, sources := range ecosystemMap {
+		ecosystemResult := HTMLEcosystemResult{
+			Ecosystem: ecosystem,
+			Sources:   sources,
+		}
+
+		if isOSImage(ecosystem) {
+			osResults = append(osResults, ecosystemResult)
+		} else {
+			ecosystemResults = append(ecosystemResults, ecosystemResult)
+		}
+	}
+
+	// Sort ecosystemResults to ensure consistent output
+	sort.Slice(ecosystemResults, func(i, j int) bool {
+		return ecosystemResults[i].Ecosystem < ecosystemResults[j].Ecosystem
+	})
+
+	ecosystemResults = append(ecosystemResults, osResults...)
+
+	return HTMLResult{
+		EcosystemResults: ecosystemResults,
+		HTMLVulnCount:    resultCount,
+	}
+}
+
+func updateCount(original *HTMLVulnCount, newAdded *HTMLVulnCount) {
+	original.Critical += newAdded.Critical
+	original.High += newAdded.High
+	original.Medium += newAdded.Medium
+	original.Low += newAdded.Low
+	original.Unknown += newAdded.Unknown
+	original.Called += newAdded.Called
+	original.Uncalled += newAdded.Uncalled
+	original.Fixed += newAdded.Fixed
+	original.UnFixed += newAdded.UnFixed
+}
+
+func addCount(resultCount *HTMLVulnCount, typeName string) {
+	switch typeName {
+	case "CRITICAL":
+		resultCount.Critical += 1
+	case "HIGH":
+		resultCount.High += 1
+	case "MEDIUM":
+		resultCount.Medium += 1
+	case "LOW":
+		resultCount.Low += 1
+	case "UNKNOWN":
+		resultCount.Unknown += 1
+	}
+}
+
+func isOSImage(ecosystem string) bool {
+	for _, image := range baseImages {
+		if strings.HasPrefix(ecosystem, image) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// uniqueIndex creates a function that generates unique indices for HTML elements.
+// It takes an integer pointer as input and increments the integer's value each time the
+// returned function is called. This ensures that each call to the returned function
+// produces a different index, even when called concurrently from multiple goroutines.
+func uniqueIndex(index *int) func() int {
+	return func() int {
+		*index += 1
+		return *index
+	}
+}
+
+// getFixVersion returns the lowest fixed version for a given package and
+// its current installed version, considering the affected ranges. If no fix is
+// available, it returns "No fix available".
+func getFixVersion(allAffected []models.Affected, installedVersion string, installedPackage string, ecosystem models.Ecosystem) string {
```

Comment:
Personally I would prefer to return a bool to indicate whether a fix is available plus the string, however I also understand we want to pass the fix version to the string field so we have to make the return value a string...

Reply:
I think we will add a centralized fix version function in the future. So I'll leave the current implementation as is for now and apply this suggestion when creating the new function.